New Roboticist and RD traitor mech: Justice.

Xackii opened the pull request:
## About The Pull Request
Video:
https://youtu.be/QwJfO_lbLK0
Text:
Add new traitor mech Justice as Roboticist and RD job restricted item:
Cost 20 tc
Health 175
Damage 60
Only melee attacks(can't attach range weapons(can't attach anything in his arms at all))
Block chanse 60
Speed - in safty mode 2.5 in combat mode 4.5
Attack effects:
   In addition to damage, it has a 50 percent chance of cutting off an arm or leg (does not put you to sleep or give a knockdown 
  effect when hit like regular combat mechs do).
    If a target in crit or stun makes a finishing attack instead of a regular attack, the mech says an epic phrase and after a second cuts off the target's head by dashing through the target.

Special actions(can't be used if mech in safty mode):

Invisability - Makes the mech invisible for 20 seconds. If the mech is pushed or collides with someone or if the pilot presses the invisibility button again, the mech will no longer be invisible and will go into a 5 second cooldown (the same will happen after 20 seconds). If you attack in invisibility, then instead of a normal attack the mech will make an AoE 3x3 attack 1 second after pressing (the attack will be accompanied by a sound and visual effect and for a second the mech will not be able to move, which gives enemies time to run away from the attack). Upon completion AoE attack, everyone( except peoples in crit and in stunned condition) in a 3x3 radius will receive 35 damage and a 25 chance of having one of their limbs cut off.

Charge Attack - Mech makes a dash towards the cursor at a distance of up to 7 cells, breaking all obstacles in its path. If the obstacle is a living person (and he is not in crit and is not stunned) then he will receive 35 damage and a 25 chance of cutting off a limb. Cooldown 5 Seconds.

Also made the sounds of mech melee attacks to vars so that it would be more convenient to change them.
Also fix a runtime error where if you set something in the mech's max_equip_by_category to 0, the mech UI panel doesn't load.

## Why It's Good For The Game

Adds a tritor item unique to the roboticist and rd (similar to chap his grace or a mime mech) with fun and interesting mechanics. This is a new mech with new unique abilities that turns the battle with you into a boss fight. 

:cl:
add: New traitor mech: Justice. For Roboticist and RD.
/:cl:

Xackii committed:
Justice

Jacquerel commented:
I will be honest I don't understand how any of this concept is relevant to lawyers

Rex9001 commented:
Pretty cool melee mech, but why is it a lawyer traitor item?

Xackii commented:
In general, the concept is something like a fighter for justice. When corrupt officers don't listen to the law, a lawyer should show them where they belong

mc-oofert commented:
sanabi

Jacquerel commented:
It isn't even wielding a gavel...

Xackii committed:
capital letters

paganiy commented:
— Ah ah. HoS! Do you read me?
— Yep. Loud and clear.
— HoS! HoS! It's so cool! There's a **giant mech** in the brig!
— A mech?
— Yes! **A giant mech!** It's so cool! Did you put this here?
— I... have never seen that thing before.
— Oh, it's says something here. **Sa... Nabi...** What's a **Sanabi**?
— Officer... **Listen to head**. **Get out of the brig right now**. **RIGHT NOW!**
— Huh? What? I can't hear you. Teehee. HoS! This mech is movi!@#%
— **NOW GET OUT OF THE BRIG**. Do you hear me? Hello?

Xackii commented:
> — Ah ah. HoS! Do you read me? — Yep. Loud and clear. — HoS! HoS! It's so cool! There's a **giant mech** in the brig! — A mech? — Yes! **A giant mech!** It's so cool! Did you put this here? — I... have never seen that thing before. — Oh, it's says something here. **Sa... Nabi...** What's a **Sanabi**? — Officer... **Listen to head**. **Get out of the brig right now**. **RIGHT NOW!** — Huh? What? I can't hear you. Teehee. HoS! This mech is movi!@#% — **NOW GET OUT OF THE BRIG**. Do you hear me? Hello?


students

Profakos commented:
Invisibility is already taken up by the Recitence... and Justice isn't invisible, its blind!

Admiral-Gigaton commented:
this is really funny, giving a mech to the lawyer of all things, but...

how in the hell does this connect to a LAWYER? sure i get the "IM BATMAN, JUSTICE" part of the gimmick, but i mean... melee only? charging? high delimb? going invis?? fun cool mechanics but again how does this connect to the lawyer, and not like the CHEF, being a melee centric combatant (if in combat, anyways) having CQC and butcher knives.

still, funny giving a mech to a lawyer of all jobs, just doesnt make sense

Jacquerel commented:
I will be honest I'm struggling to imagine a world where we put this into the game.
I'm extremely iffy on putting any more mechs in the uplink for _anyone_.
I cannot see how we would give one to the lawyer.

IDK maybe you can find another maintainer to argue your corner who is seeing something I am missing but, I would not hold your breath.

Xackii commented:
> I will be honest I'm struggling to imagine a world where we put this into the game. I'm extremely iffy on putting any more mechs in the uplink for _anyone_. I cannot see how we would give one to the lawyer.
> 
> I can see a world where we give some kind of item to laywers but I can't imagine how it would be this one. This item has basically nothing to do with laywers at all. It's a huge non-sequitur.
> 
> IDK maybe you can find another maintainer to argue your corner who is seeing something I am missing but, I would not hold your breath.

This mech is from the game sanabi (something like that). I don’t know who else might be suitable for a mech that, according to the lore, should fight for justice (an assistant?). Initially, I thought about adding this mech to nuki ops, but I wasn’t sure if this mech was suitable for them too.

Jacquerel commented:
Telling me that this is a reference to a video game unfortunately is not a very appealing argument to me; I don't care about the lore of some other video game, I only care about what you are adding to space station 13.

Admiral-Gigaton commented:
> ss13 is the embodiment of references from other games

bruh their argument still stands, you need some real reasoning for this to work

Xackii commented:
> > ss13 is the embodiment of references from other games
> 
> bruh their argument still stands, you need some real reasoning for this to work

I don’t know what arguments can be inserted here. The complaint is that he doesn't think mech looks like mech for a lawyer, and neither do you. But what should I do about it? The reason why they want to close it is that, according to the maintainer, the mech does not look like the mech of a lawyer. They didn’t tell me what I should do to prevent the PR from being closed. What should I answer? So I think mech is suitable for a lawyer? This is what I think, and not the maintainer who thinks so. If the maintainer has ideas that I can make to make the mech more suitable for how he sees the lawyer's mech, then I'm listening.

PunishedHideo commented:
Sanabi deniers should perish

Xackii commented:
if not a lawyer, then i you can add this mech to ops. 
all their mechs are an improved version of the mechs that are on the station and this mech provides unique mechanics

Admiral-Gigaton commented:
> if not a lawyer, then i you can add this mech to ops. all their mechs are an improved version of the mechs that are on the station and this mech provides unique mechanics

This is true, ops don't have much else for mechs besides the meta, that being the mauler. Even then, face to face combat tends to do better (speed and shitty mech wounds that turn it into a spinning heap of shit that suffocates and burns you) and it would probably fit better for ops than... the lawyer

tralezab commented:
Alright settle down fellas... it's clear there is a large fundamental issue with how you get this mech (mechs just do not make good job specific traitor items, in fact mechs do not make good traitor items period, i'm with jac on that) but otherwise the mech itself considering nothing else is very cool and so i ask: is there another implementation for how you get the mech that would be more agreeable? some kind of unique traitor sabotage like how antagonists can hack meteor satellites to get the dark matteor? that's how i lock cool things behind hard work 💪

tralezab commented:
> The Shogun is a lawyer of his time

this post goes insanely hard for a github comment and i think it properly stresses my point that we should not throw the baby out with the bathwater

Xackii commented:
> Alright settle down fellas... it's clear there is a large fundamental issue with how you get this mech (mechs just do not make good job specific traitor items, in fact mechs do not make good traitor items period, i'm with jac on that) but otherwise the mech itself considering nothing else is very cool and so i ask: is there another implementation for how you get the mech that would be more agreeable? some kind of unique traitor sabotage like how antagonists can hack meteor satellites to get the dark matteor? that's how i lock cool things behind hard work 💪


I just don’t know where. Meteor shields and dark meteor sound logical. how to make it logical for mech.

Sakamoto4ka commented:
> > The Shogun is a lawyer of his time
> 
> no one cares

you just need to make this pr after someone added judge mech with big hammer. Its simple

Xackii commented:
> > > The Shogun is a lawyer of his time
> > 
> > 
> > no one cares
> 
> you just need to make this pr after someone added judge mech with big hammer. Its simple


then my pr will not be relevant and they will tell me that I don’t understand anything about art

Rex9001 commented:
What if you lock the parts behind emagging an exofab?

Xackii commented:
> What if you lock the parts behind emagging an exofab?

it's possible

Profakos commented:
The mech should also have different sprites, direct references to other games do not go down well anymore, and they neither should.

Xackii commented:
> The mech should also have different sprites, direct references to other games do not go down well anymore, and they neither should.

Well, this is not an exact copy of the mech, but that’s how I tried to transfer this idea to SS13

Jackal-boop commented:
how is this related to the lawyer in any way. job tot items should be in line with their occupation. EG doctors get evil surguries.

SomeoneYouProbablyKnow commented:
Everyone's focused on how it's a mech and how it's not lawyerly but i'm just curious if it's any good or any fun.

As is you're gonna cut off a few heads and get lasered to death, yea you got block chance and invisibility but what are you going to do about to dudes running around you shooting you? You just die and there go your 20 TC.
And if you don't die... you just stab people untill they die. Just a more complicated version of a sword.

It does LOOK cool as heck even if i have no idea what the reference is

projectkepler-ru commented:
ok first and foremost
you're talented
and this is cool
but how is this a lawyer thing???

moocowswag commented:
It should be liberty prime mech

Xackii committed:
not_for_you_lawyer

Xackii commented:
> I will be honest I don't understand how any of this concept is relevant to lawyers

Ok i made it roboticist and rd job restricted item instead of lawyer

Jacquerel commented:
I still don't know if I want another "buy a full mech" in the uplink but I'll remove the tag in case another maintainer is interested in this PR.

Xackii commented:
I'm still open to ideas on where or how to put this mech. The idea to add it to the lawyer was also so that this mech would not be bought every round

projectkepler-ru commented:
> I'm still open to ideas on where or how to put this mech. The idea to add it to the lawyer was also so that this mech would not be bought every round

perfectly reasonable 
dont get me wrong this is really cool and well coded, it was ust seemingly abit excessive at first lol

TheNobodyofimportance commented:
How about this for building the mech, make it an upgrade like the ripley MK.2 that is applied to a durand/gygax/whatever. Thus there needs to be a measure of investment and work into gaining the mech rather than just popping it down inside the clown room [He slipped me :(]

Whatever mech you need to build is at your discretion of course but this seems like an elegant way to build it.

Xackii commented:
> How about this for building the mech: make it an upgrade like the ripley MK.2 that is applied to a durand/gygax/whatever. Thus there needs to be a measure of investment and work into gaining the mech rather than just popping it down inside the clown room [He slipped me :(]
> 
> Whatever mech you need to build is at your discretion of course but this seems like an elegant way to build it.

like some kind of tritor thing that attaches to an existing mech and turns it into a new one?

scrunklershrungle932 commented:
> Alright settle down fellas... it's clear there is a large fundamental issue with how you get this mech (mechs just do not make good job specific traitor items, in fact mechs do not make good traitor items period, i'm with jac on that) but otherwise the mech itself considering nothing else is very cool and so i ask: is there another implementation for how you get the mech that would be more agreeable? some kind of unique traitor sabotage like how antagonists can hack meteor satellites to get the dark matteor? that's how i lock cool things behind hard work 💪

Not related to the mech or anything, but I didn't know hacking meteor satellites is a thing that could be done.  What's dark matteor or the dark matteor?

Blacklist897 commented:
> > Alright settle down fellas... it's clear there is a large fundamental issue with how you get this mech (mechs just do not make good job specific traitor items, in fact mechs do not make good traitor items period, i'm with jac on that) but otherwise the mech itself considering nothing else is very cool and so i ask: is there another implementation for how you get the mech that would be more agreeable? some kind of unique traitor sabotage like how antagonists can hack meteor satellites to get the dark matteor? that's how i lock cool things behind hard work 💪
> 
> Not related to the mech or anything, but I didn't know hacking meteor satellites is a thing that could be done. What's dark matteor or the dark matteor?

its from a traitor final objective or if 8 meteor sats gets emagged, its basically a singlo that chases the closest living thing

Admiral-Gigaton commented:
FINALLY, this makes more sense as a PR

awesome, great job, that was really the only reason people shat on it, love the idea, great execution, godspeed

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+			new /obj/effect/temp_visual/mech_sparks(get_turf(src))
+			playsound(src, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+			return
+	. = ..()
```
```suggestion
	return ..()
```

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+/datum/action/vehicle/sealed/mecha/invisibility
+	name = "Invisibility"
+	button_icon_state = "mech_stealth_off"
+	var/on = FALSE
+	var/charge = TRUE
+	var/start_attack = FALSE
+	var/stealth_pre_attack_sound = 'sound/mecha/mech_stealth_pre_attack.ogg'
+	var/stealth_attack_sound = 'sound/mecha/mech_stealth_attack.ogg'
```
autodoc comments

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+		for(var/who_inside in chassis.occupants)
+			var/mob/living/occupant = who_inside
+			var/datum/action/vehicle/sealed/mecha/charge_attack/charge_action = LAZYACCESSASSOC(chassis.occupant_actions, occupant, /datum/action/vehicle/sealed/mecha/charge_attack)
+			if(charge_action)
+				if(charge_action.on)
```
`if(charge_action?.on)`

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+		UnregisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK)
+	build_all_button_icons()
+
+/datum/action/vehicle/sealed/mecha/invisibility/proc/end_stealth()
```
autodoc

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+		owner.balloon_alert(owner, "invisability is over")
+		Trigger()
+
+/datum/action/vehicle/sealed/mecha/invisibility/proc/stealth_attack_aoe(datum/source, mob/living/pilot, atom/target, on_cooldown, is_adjacent)
```
autodoc

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+	addtimer(CALLBACK(src, PROC_REF(attack_in_aoe), pilot), 1 SECONDS)
+	return TRUE
+
+/datum/action/vehicle/sealed/mecha/invisibility/proc/attack_in_aoe(mob/living/pilot)
```
autodoc

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+	build_all_button_icons()
+	addtimer(CALLBACK(src, PROC_REF(charge)), 5 SECONDS)
+
+/datum/action/vehicle/sealed/mecha/invisibility/proc/charge()
```
autodoc

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+/datum/action/vehicle/sealed/mecha/charge_attack
+	name = "Charge Attack"
+	button_icon_state = "mech_charge_off"
+	var/on = FALSE
+	var/charge = TRUE
+	var/max_charge_range = 7
+	var/charge_attack_sound = 'sound/mecha/mech_charge_attack.ogg'
```
comments

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+		UnregisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK)
+	build_all_button_icons()
+
+/datum/action/vehicle/sealed/mecha/charge_attack/proc/click_try_charge(datum/source, mob/living/pilot, atom/target, on_cooldown, is_adjacent)
```
same

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+	addtimer(CALLBACK(src, PROC_REF(charge)), 5 SECONDS)
+	return TRUE
+
+/datum/action/vehicle/sealed/mecha/charge_attack/proc/charge()
```
comments

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+	forceMove(finish_turf)
+
+/obj/vehicle/sealed/mecha/justice/melee_attack_effect(mob/living/victim, heavy)
+	if(heavy)
+		var/obj/item/bodypart/cut_bodypart = victim.get_bodypart(pick(BODY_ZONE_R_ARM, BODY_ZONE_R_LEG, BODY_ZONE_L_ARM, BODY_ZONE_L_LEG))
+		if(prob(DISMEMBER_CUALITY_HIGTH))
+			cut_bodypart.dismember(BRUTE)
+	else
+		victim.Knockdown(40)
```
pls just early return !heavy, else is bad

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+
+/obj/vehicle/sealed/mecha/justice/process(seconds_per_tick)
+	. = ..()
+	for(var/who_inside in occupants)
+		var/mob/living/occupant = who_inside
```
`for(var/mob/living/occupant in occupants)`?

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+
+	if(isliving(target))
+		var/mob/living/live_or_dead = target
+		if((live_or_dead.stat >= UNCONSCIOUS && live_or_dead.stat < DEAD) || live_or_dead.getStaminaLoss() >= 100)
```
early return this

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+	if(isliving(target))
+		var/mob/living/live_or_dead = target
+		if((live_or_dead.stat >= UNCONSCIOUS && live_or_dead.stat < DEAD) || live_or_dead.getStaminaLoss() >= 100)
+			if(charge_action) // We don't do fatality if try to use charge or stealth attack
+				if(charge_action.on)
```
`charge_action?.on`

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+			playsound(src, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+
+/obj/vehicle/sealed/mecha/justice/take_damage(damage_amount, damage_type, damage_flag, sound_effect, attack_dir, armour_penetration)
+	if(stealth_action)
+		if(stealth_action.on)
```
`stealth_action?.on`

jlsnow301 commented on code/modules/vehicles/mecha/combat/justice.dm:
```diff
@@ -0,0 +1,394 @@
+			return TRUE
+	return FALSE
+
+/datum/action/vehicle/sealed/mecha/charge_attack/proc/charge_attack(mob/living/charger, turf/target)
```
autodoc comment

jlsnow301 commented on code/modules/vehicles/mecha/mecha_ui.dm:
```diff
@@ -111,36 +111,37 @@
+						"slot" = category
+					))
+					if(ui_selected_module_index == module_index)
+						ui_selected_module_index = null
+				else
```
just continue if you can so we don't have to else block this

jlsnow301 reviewed (changes requested):
Sprites are sick, need some tlc on the code though,
[Autodoc](https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md)
[Early returns](https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-early-returns)
&& optional chaining will help your code quite a bit.

Xackii committed:
autodoc_autocode_automerge_autoclose

Xackii committed:
effectsdmi

Xackii commented:
icons moment

Xackii committed:
maybethis

Xackii commented:
> Sprites are sick, need some tlc on the code though, [Autodoc](https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md) [Early returns](https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-early-returns) && optional chaining will help your code quite a bit.

A strange problem with icons associated with the new pr for raptors which added a couple of icons to the effects folder. I updated the build and added my icon again, but the merge conflict is not removed. Do I need to reopen pr or what? Also make autodoc and early returns

Xackii committed:
Merge branch 'master' into ronny

Xackii commented:
ok, AnturK helped with solving the problem

Xackii committed:
Merge branch 'master' into ronny

Xackii committed:
Merge branch 'master' into ronny

Xackii commented:
3 weeks later

Xackii committed:
Merge branch 'master' into ronny

Xackii committed:
Merge branch 'master' into ronny

Xackii committed:
sprites

Xackii committed:
Merge branch 'master' into ronny

Xackii committed:
Merge branch 'master' into ronny

Xackii committed:
Merge branch 'master' into ronny

Deultus commented:
The next enhancement of stun batons will be integrated 10 times faster than this

Orchian commented:
Please tell me that decapitation didn't delete the brain... Other than that, pretty neat idea.

LemonInTheDark commented:
@tralezab @jlsnow301 do one of you want this pr as is? (I ask because of jacqu's concerns and the inaction)

jlsnow301 commented:
> @tralezab @jlsnow301 do one of you want this pr as is? (I ask because of jacqu's concerns and the inaction)

i'm a bit biased against mechs, probably not the best person to ask. yes, the sprites are great, but i just don't like mechs enough to want to add anything to them

LemonInTheDark commented:
> i'm a bit biased against mechs, probably not the best person to ask. yes, the sprites are great, but i just don't like mechs enough to want to add anything to them

~~why did you review it then ahhhh~~

jlsnow301 commented:
> ~why did you review it then ahhhh~

Valid, my apologies, I should have stated directly. I would not block this PR but I have no personal interest in changing mechs (outside of removal)

Xackii committed:
Merge branch 'master' into ronny

Jacquerel commented:
Despite the tag being removed it doesn't seem like there's any interest in merging this and I would frankly usually just rather science staff build mechs instead of buying a fully-made one from an uplink. You can reopen this if you find a maintainer who actually really wants it in the game.

LemonInTheDark commented:
If you are interested in perusing this further, I would recommend adding a more interesting source for the mech. We don't want people (especially science) just buying something they could make. @tralezab (armhulen on discord) has volenteered to have opinions on this topic, you should reach out to him for more info

![image](https://github.com/tgstation/tgstation/assets/58055496/1270e3d4-dd2f-47c1-ae52-87a90932d1c6)

Xackii commented:
maybe add mech to nukers?

LemonInTheDark commented:
don't they already have two? would it fill an interesting role?

Xackii commented:
> don't they already have two? would it fill an interesting role?

those two are heavy and long-range. This one is more stealth and melee

Orchian commented:
It doesn't really seem like a mech for standard crew is the issue I see with having it be built as opposed to 'bought.' Maybe having a parts kit intended to be used on a different, existing chassis could be a feasible solution if the issue is simply, not wanting the mech to appear out of thin air?

It seems worthwhile to me to follow up with Tralezab on this one, at least to understand what their gripes are with the current idea in better detail.
